odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Functions for folders and files."""
+
+from __future__ import print_function
+import os
+import numpy as np
+import odl
+from scipy.signal import convolve2d as scipy_convolve2d
+
+__author__ = "Matthias J. Ehrhardt"
+__copyright__ = "Copyright 2016, University of Cambridge"
+__all__ = ('exists', 'mkdir', 'TV', 'TV_NonNegative', 'bregman',
+           'divide_1Darray_equally', 'Blur', 'HuberL1',
+           'KullbackLeiblerSmooth')
+
+
+def exists(path):
+    return os.path.exists(path)
+
+
+def mkdir(path):
+    if not os.path.exists(path):
+        os.makedirs(path)
+
+
+def bregman(f, v, r):
+    return (odl.solvers.FunctionalQuadraticPerturb(f, linear_term=-r)
+            - f(v) + r.inner(v))
+
+
+def divide_1Darray_equally(ind, n_subsets):
+    """
+    Divide an array into equal chunks to be used for instance in OSEM.
+
+    Parameters
+    ----------
+    ind : ndarray
+        input array
+    nsubsets : int
+        number of subsets to be divided into
+
+    Returns
+    -------
+    subset2ind : list
+        list of indices for each subset
+    ind2subset : list
+        list of subsets for each index
+    """
+
+    n_ind = len(ind)
+    subset2ind = []
+
+    for i in range(n_subsets):
+        subset2ind.append([])
+        for j in range(i, n_ind, n_subsets):
+            subset2ind[i].append(ind[j])
+
+    ind2subset = []
+    for i in range(n_ind):
+        ind2subset.append([])
+
+    for i in range(n_subsets):
+        for j in subset2ind[i]:
+            ind2subset[j].append(i)
+
+    return (subset2ind, ind2subset)
```

Comment:
What is "a" here?

Reply:
That was a copy-paste error, I edited it away.

Reply:
The point of this function is to create a mapping from subsets to indices and from indices to subsets. In general a subset will contain many indices and an index may be in many subsets. I am not sure your function provides the latter in general. In any ideas how to achieve this like my code but as elegant as yours?

Reply:
> The point of this function is to create a mapping from subsets to indices and from indices to subsets. In general a subset will contain many indices and an index may be in many subsets. I am not sure your function provides the latter in general. In any ideas how to achieve this like my code but as elegant as yours?

In that case, I would make it two functions, one doing the partitioning and one for the inverse. A more generic API for splitting would be something like
```python
def partition_1d(arr, slices):
    return tuple(arr[slc] for slc in slices)


def partition_equally_1d(arr, n_parts, order='block'):
    if order == 'block':
        stride = int(np.ceil(arr.size / n_parts))
        slc_list = [slice(i * stride, (i + 1) * stride) for i in range(n_parts)]
    elif order == 'interlaced':
        slc_list = [slice(i, arr.size, n_parts) for i in range(n_parts)]
    else:
        raise ValueError

    return partition_1d(arr, slc_list)
```
You could go more fancy than that, of course. Also "equally" is so-so here since for `ind = np.arange(10)` you get
```python
>>> partition_equally_1d(ind, 4, order='block')
(array([0, 1, 2]), array([3, 4, 5]), array([6, 7, 8]), array([9]))
>>> partition_equally_1d(ind, 4, order='interlaced')
(array([0, 4, 8]), array([1, 5, 9]), array([2, 6]), array([3, 7]))
```
For the inverse I can't think of an elegant way other than the naive implementation. The problem seems to me that the result is a list of lists with varying sizes, so it's a bit hard to use a Numpy array for that.